### ISSUE TITLE
Inconsistent error CS8120: The switch case has already been handled by a previous case.

### ISSUE THREAD

**Issue by TessenR** [dotnet/roslyn#20210]
**Version Used**: 
VisualStudio.15.Preview/15.3.0-Preview+26510.0

**Steps to Reproduce**:

Consider the following code:

```cs
class Program
{
  public static void Main()
  {
    switch (default(object)) // OK
    {
      case bool _:
      case true:
      case false:
        break;
    }

    switch (new object()) // errors
    {
      case bool _:
      case true:
      case false:
        break;
    }

    switch ((object)null) // OK
    {
      case bool _:
      case true:
      case false:
        break;
    }

    switch ((bool?)null) // OK
    {
      case bool _:
      case true:
      case false:
        break;
    }

    switch (default(bool?)) // errors
    {
      case bool _:
      case true:
      case false:
        break;
    }

    switch (default(bool)) // errors
    {
      case bool _:
      case true:
      case false:
        break;
    }
  }
}
```

**Expected Behavior**:
```
error CS8120: The switch case has already been handled by a previous case.
```
is reported either for each switch statement in the program or for neither of them.

**Actual Behavior**:
the error is reported only for some of switch statements in the program.
It is reported for switches over `new object()`, `default(bool?)` and `default(bool)` but not for switches over `default(object)`, `(object)null`, `(bool?)null`.

**Comment by jaredpar** [comment]
Believe this is the representative case to examine

``` c#
 switch ((object)null) // OK
    {
      case bool _:
      case true:
      case false:
        break;
    }
```

All of the other cases essentially boil down to this core example: a constant expression `null` given an explicit type. 

Unsure if this is a bug or not.  Will let @gafter comment.
